### PR TITLE
CSI 4969 prepare before controller util files

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: volume-group-operator/main.go [hunk 1]
```diff
@@ -79,18 +85,22 @@ func main() {
 	})
 	exitWithError(err, "unable to start manager")
 
-	log := logr.Logger{}
+	log := ctrl.Log.WithName("controllers").WithName("VolumeGroup")
 	grpcClientInstance, err := getControllerGrpcClient(cfg, log)
 	exitWithError(err, err.Error())
 
+	controllerUtils := utils.ControllerUtils{
```
Comment (matancarmeli7) [inline volume-group-operator/main.go:92]: why do we want to do it?

**Review by matancarmeli7 — COMMENTED** [review]
if there are comments that will take you time to handle you can open a ticket (except the list volumeGroupClass comment)

File: volume-group-operator/controllers/utils/controller_utils.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package utils
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/controller_utils.go:1]: we have CSI-4880 for it..

File: volume-group-operator/controllers/utils/volumegroupclass.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+// GetVolumeGroupClass get volume group class object from the subjected namespace and return the same.
+func (r *ControllerUtils) GetVolumeGroupClass(logger logr.Logger, vgcName string) (*volumegroupv1.VolumeGroupClass, error) {
+	vgcObjl := &volumegroupv1.VolumeGroupClassList{}
+	err := r.Client.List(context.TODO(), vgcObjl, &client.ListOptions{Raw: &metav1.ListOptions{FieldSelector: fmt.Sprintf("metadata.name=%s", vgcName)}})
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/volumegroupclass.go]: done

File: volume-group-operator/controllers/utils/volumegroupclass.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2021.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/volumegroupclass.go]: done

File: volume-group-operator/main.go (hunk 1 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go:92]: so they would have an initiated `Client`

File: volume-group-operator/controllers/utils/finalizers.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+				VolumeGroupFinalizer, vg.Namespace, vg.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// AddFinalizerToVGC adds the VG finalizer on the VolumeGroupContent instance.
+func (r *ControllerUtils) AddFinalizerToVGC(logger logr.Logger, vgc *volumegroupv1.VolumeGroupContent,
+) error {
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go:52]: what line?

File: volume-group-operator/controllers/utils/finalizers.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	if !Contains(vgc.ObjectMeta.Finalizers, volumeGroupContentFinalizer) {
+		logger.Info("adding finalizer to volumeGroupContent object", "Finalizer", volumeGroupContentFinalizer)
+		vgc.ObjectMeta.Finalizers = append(vgc.ObjectMeta.Finalizers, volumeGroupContentFinalizer)
+		if err := r.Client.Update(context.TODO(), vgc); err != nil {
+			return fmt.Errorf("failed to add finalizer (%s) to volumeGroupContent resource"+
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go:57]: we have for messages already, do we need another?

[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/finalizers.go:57]: the messages ticket will include creating something similar to [this](https://github.com/IBM/ibm-block-csi-driver/blob/develop/node/pkg/driver/device_connectivity/errors.go)?

File: volume-group-operator/controllers/utils/finalizers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/finalizers.go:52]: image

[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/finalizers.go:52]: that it will nicer

File: volume-group-operator/main.go (hunk 1 above)
Comment (matancarmeli7) [reply volume-group-operator/main.go:92]: but you already do it [here](https://github.com/IBM/ibm-block-csi-operator/blob/d10932232af81a6ff670236a821d15d81fb6d6f7/volume-group-operator/main.go#L97), no?

File: volume-group-operator/controllers/utils/finalizers.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	VolumeGroupFinalizer        = "volumegroup.storage.ibm.io"
+	volumeGroupContentFinalizer = "volumegroup.storage.ibm.io/vgc-protection"
+	pvcVolumeGroupFinalizer     = "volumegroup.storage.ibm.io/pvc-protection"
+)
+
+// AddFinalizerToVG adds the VG finalizer on the VolumeGroup instance.
+func (r *ControllerUtils) AddFinalizerToVG(logger logr.Logger, vg *volumegroupv1.VolumeGroup,
+) error {
+	if !Contains(vg.ObjectMeta.Finalizers, VolumeGroupFinalizer) {
+		logger.Info("adding finalizer to VolumeGroup object", "Finalizer", VolumeGroupFinalizer)
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go:38]: basically the name and namespace of the Reconciled object always in the log but I will add them to the other objects

File: volume-group-operator/controllers/utils/finalizers.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	VolumeGroupFinalizer        = "volumegroup.storage.ibm.io"
+	volumeGroupContentFinalizer = "volumegroup.storage.ibm.io/vgc-protection"
+	pvcVolumeGroupFinalizer     = "volumegroup.storage.ibm.io/pvc-protection"
+)
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: done

File: volume-group-operator/controllers/utils/finalizers.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	"context"
+	"fmt"
+	volumegroupv1 "github.com/IBM/volume-group-operator/api/v1"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	VolumeGroupFinalizer        = "volumegroup.storage.ibm.io"
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: done

File: volume-group-operator/controllers/utils/finalizers.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	volumeGroupContentFinalizer = "volumegroup.storage.ibm.io/vgc-protection"
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: done

File: volume-group-operator/controllers/utils/finalizers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: done

File: volume-group-operator/controllers/utils/finalizers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: it looks unnecessary and complex, why is it good?

File: volume-group-operator/controllers/utils/finalizers.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go:28]: done

File: volume-group-operator/controllers/utils/finalizers.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+func (r *ControllerUtils) AddFinalizerToVG(logger logr.Logger, vg *volumegroupv1.VolumeGroup,
+) error {
+	if !Contains(vg.ObjectMeta.Finalizers, VolumeGroupFinalizer) {
+		logger.Info("adding finalizer to VolumeGroup object", "Finalizer", VolumeGroupFinalizer)
+		vg.ObjectMeta.Finalizers = append(vg.ObjectMeta.Finalizers, VolumeGroupFinalizer)
+		if err := r.Client.Update(context.TODO(), vg); err != nil {
+			return fmt.Errorf("failed to add finalizer (%s) to volumeGroup resource"+
+				" (%s/%s) %w",
+				VolumeGroupFinalizer, vg.Namespace, vg.Name, err)
+		}
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/finalizers.go]: done

File: volume-group-operator/controllers/utils/secret.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	"context"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// GetSecret retrieves the secrets based on name and namespace input.
+func (r *ControllerUtils) GetSecret(logger logr.Logger, name, namespace string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/secret.go]: done

File: volume-group-operator/controllers/utils/suite_test.go [hunk 13]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package controllers
+package utils
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/suite_test.go]: done

File: volume-group-operator/controllers/utils/volumegroupcontent.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		logger.Error(err, "VolumeGroupContent creation failed", "VolumeGroupContent Name")
+		return err
+	}
+
+	return nil
+}
+
+// GenerateVolumeGroupContent create an VolumeGroupContent object.
+func (r *ControllerUtils) GenerateVolumeGroupContent(instance *volumegroupv1.VolumeGroup, vgcObj *volumegroupv1.VolumeGroupClass, resp *volumegroup.Response, secretName string, secretNamespace string, groupCreationTime *metav1.Time, ready *bool) *volumegroupv1.VolumeGroupContent {
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/utils/volumegroupcontent.go]: done

File: volume-group-operator/controllers/volumegroup/volume_group.go [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+// CommonRequestParameters holds the common parameters across volume group operations.
+type CommonRequestParameters struct {
+	Name          string
+	VolumeGroupID string
+	VolumeIds     []string
+	Parameters    map[string]string
+	Secrets       map[string]string
+	VolumeGroup   client.VolumeGroup
+}
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/volumegroup/volume_group.go:44]: its types that are relevant only to the file..

File: volume-group-operator/main.go (hunk 1 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go:90]: but they are not the same package, do you want to pass the client to each method like the logger?

Comment (ArbelNathan) [reply volume-group-operator/main.go]: done

File: volume-group-operator/controllers/utils/controller_utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/controller_utils.go:1]: it's two seconds work, but if you want to wait for the ticket than resolve this comment

File: volume-group-operator/controllers/utils/volumegroupcontent.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/volumegroupcontent.go]: I still see it like this
image

File: volume-group-operator/controllers/volumegroup/volume_group.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/volumegroup/volume_group.go:44]: I meant a type file in this package, if you don't think it's necessary you can resolve this comment

[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/volumegroup/volume_group.go:44]: and if only this file is using them, why are they with upper case letters at their beginning?

[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/volumegroup/volume_group.go]: I believe there is a use, I split them into separate files

File: volume-group-operator/controllers/utils/finalizers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ELENAGER) [reply volume-group-operator/controllers/utils/finalizers.go]: According to Occam's razor principle, "entities should not be multiplied beyond necessity". I think, one messages ticket is enough.

[the remark this answers is not in the document]
Comment (ELENAGER) [reply volume-group-operator/controllers/utils/finalizers.go]: Sorry, I don't understand the subject of the discussion. One day we will have the messages file. We will discuss it later when writing it.

[the remark this answers is not in the document]
Comment (ELENAGER) [reply volume-group-operator/controllers/utils/finalizers.go]: Frankly - I think it's unnecessary detalization and overload. I can happily live without it

[the remark this answers is not in the document]
Comment (matancarmeli7) [reply volume-group-operator/controllers/utils/finalizers.go]: ok resolving this comment

**Review by matancarmeli7 — APPROVED** [review]
LGTM

**Review by ELENAGER — APPROVED** [review]
LGTM

**Pull request merged**